plugins/ccgx/fu-ccgx-pure-hid-device.c

```diff
@@ -0,0 +1,472 @@
+	guint version = 0;
+	g_autofree gchar *bl_ver = NULL;
+	g_autoptr(GByteArray) st_info = NULL;
+
+	if (!fu_hid_device_get_report(FU_HID_DEVICE(self),
+				      buf[0],
+				      buf,
+				      sizeof(buf),
+				      FU_CCGX_PURE_HID_DEVICE_TIMEOUT,
+				      FU_HID_DEVICE_FLAG_IS_FEATURE,
+				      error))
+		return FALSE;
+
+	st_info = fu_struct_ccgx_pure_hid_fw_info_parse(buf, sizeof(buf), 0x0, error);
+	if (st_info == NULL)
+		return FALSE;
+	self->silicon_id = fu_struct_ccgx_pure_hid_fw_info_get_silicon_id(st_info);
+	self->operating_mode = fu_struct_ccgx_pure_hid_fw_info_get_operating_mode(st_info);
+
+	fu_device_remove_flag(device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+	/* set current version */
+	switch (self->operating_mode) {
+	case FU_CCGX_PURE_HID_FW_MODE_FW1:
+		version = fu_struct_ccgx_pure_hid_fw_info_get_image1_version(st_info);
+		break;
+	case FU_CCGX_PURE_HID_FW_MODE_FW2:
+		version = fu_struct_ccgx_pure_hid_fw_info_get_image2_version(st_info);
+		break;
+	case FU_CCGX_PURE_HID_FW_MODE_BOOT:
+		/* force an upgrade to any version */
+		version = 0x0;
+		fu_device_add_flag(device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+		break;
+	default:
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "unsupported mode");
+		return FALSE;
+	}
+	fu_device_set_version_raw(device, version);
+
+	/* set bootloader version */
+	fu_device_set_version_bootloader_raw(
+	    device,
+	    fu_struct_ccgx_pure_hid_fw_info_get_bl_version(st_info));
+	bl_ver = fu_version_from_uint32(fu_struct_ccgx_pure_hid_fw_info_get_bl_version(st_info),
+					fu_device_get_version_format(self));
+	fu_device_set_version_bootloader(FU_DEVICE(self), bl_ver);
+
+	/* FIXME: d4s: do we need that for querying??? */
+	if (!fu_ccgx_pure_hid_enter_flashing_mode(self, error))
+		return FALSE;
+
+	/* TODO: d4s: from wireshark for ReadVersion.exe:
+	 * > e1 04 01 00 cc cc cc cc
+	 * > e4 42 41 41 00 00 00 18
+	 */
+
+	return TRUE;
+}
+
+static gboolean
+fu_ccgx_pure_hid_device_setup(FuDevice *device, GError **error)
+{
+	FuCcgxPureHidDevice *self = FU_CCGX_PURE_HID_DEVICE(device);
+
+	/* FuUsbDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_ccgx_pure_hid_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	if (!fu_ccgx_pure_hid_magic_unlock(self, error))
+		return FALSE;
+	if (!fu_ccgx_pure_hid_ensure_fw_info(self, error))
+		return FALSE;
+
+	fu_device_add_instance_strup(device,
+				     "MODE",
+				     fu_ccgx_pure_hid_fw_mode_to_string(self->operating_mode));
+	if (!fu_device_build_instance_id(FU_DEVICE(self), error, "USB", "VID", "PID", "MODE", NULL))
+		return FALSE;
+
+	fu_device_add_instance_u16(FU_DEVICE(self), "SID", self->silicon_id);
+	if (!fu_device_build_instance_id_full(FU_DEVICE(self),
+					      FU_DEVICE_INSTANCE_FLAG_QUIRKS,
+					      error,
+					      "CCGX",
+					      "SID",
+					      NULL))
+		return FALSE;
+	g_debug("got silicon ID: 0x%04x", self->silicon_id);
```

Comment:
can this be in _to_string() instead please

Reply:
removed debug. it is already in `_to_string()`